database/ffldb/db.go

```diff
@@ -980,6 +986,19 @@ type transaction struct {
 // Enforce transaction implements the database.Tx interface.
 var _ database.Tx = (*transaction)(nil)
 
+// blockIdx2Bucket returns the new block index bucket.
+func (tx *transaction) blockIdx2Bucket() (*bucket, error) {
```

Comment:
I would suggest leaving this caching out since it slightly increases the complexity, and unless you have future plans to make more use of it, it seems like the reality is that it is only ever used to get the initial block count, to foreach the headers, and when writing new blocks to the database.  The first two only happen once, and the last one can easily lookup the bucket once outside of the write loop.

It is also per transaction, so even if more uses are added in the future, it won't be cached across transactions and not really be used then either.